### DISCUSSION TITLE
Adding a page Alert - errors and inconsistencies

### DISCUSSION THREAD

**Discussion by JoelMTaylor** [retypeapp/retype#316] [Q&A] [answered]
Hello again,

Thank you for your help with the Azure page setup. We are still trying to tweak the Azure side, but your assistance was most helpful in getting it to a functional state.

For the documentation website, we are going to have one of two alerts at the top of the page to inform the reader about whether the page and content has been reviewed:
```
!!!success
This article has been formally reviewed and approved for use.
!!!
```
or
```
!!!warning
This article has been published for your convenience. It has not yet been formally reviewed for policy compliance.
!!!
```

I have been testing the placement of the Alert, Main Heading and Metadate content (author, categories and date, mainly). During the testing, a few situations came about:
1. Author, categories and date being spread across the screen.
2. Text not displaying
3. Alerts not displaying

After further investigation, it seems that the build and deploy of the content from Retype is not working properly (somehow becoming corrupt in the process). In some instances, a build/deploy will fix or break the content displayed (even though the page does not change). I did think that it could be related to the Azure side of things, but realised that the built *.html pages are corrupted (and Azure only displays it).

I have copied some screenshots and *.html files to show as examples. They can be viewed at [https://joelmtaylor.github.io/Documentation/_example files/...](https://github.com/JoelMTaylor/Documentation/tree/main/_example%20files).

Geoff (or Fabricio), if you want, we can give you permissions to view the repo https://github.com/ccmltd/docs-corportatesystems.

**Comment by geoffreymcgill** [comment]
> Geoff (or Fabricio), if you want, we can give you permissions to view the repo https://github.com/ccmltd/docs-corportatesystems.

Please. Access to the repo should help.

**Reply by geoffreymcgill** [reply]
Seeing the `.md` source for the following page would be helpful:
https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/Author%20and%20content.png

Although being able to run the entire project locally would be the best.

**Reply by JoelMTaylor** [reply]
They should be the same as here: https://github.com/JoelMTaylor/Documentation/tree/main/CorporateSystems.

I copied the files to here to test them at a different location. There does seem to be less issues in my personal repo.

The other files would be here: https://github.com/JoelMTaylor/Documentation/tree/main/admin.

**Reply by geoffreymcgill** [reply]
It looks like the number of `categories` is not wrapping. You are assigning a lot of categories to the document, which creates a long list. 

To test, can you try commenting out about half of those categories?

```yml
categories:
  - Edumate Coordinators
  - Teaching Staff
  - Registrar
  - Office Administrators
  - Student Services
  - Fees & Finance
  # - Timetablers
  # - Curriculum Leaders
  # - Student Welfare
  # - Aides & Support
  # - Staff Management
```

I guess we just never assumed that many Categories would be used.

This is a bug, but should be an easy fix.

I'll look through the site and see if I can notice any other issues.

**Reply by geoffreymcgill** [reply]
That `<br>` at the top of the page should not be required and can be removed. 

https://github.com/JoelMTaylor/Documentation/blob/main/CorporateSystems/Multi-factor%20Authentication.md?plain=1#L29

**Reply by JoelMTaylor** [reply]
I tried the test to see if commenting out categories did anything. I didn't have any success. The examples files are:
- https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/comment%20out%20categories%20-%20MFA.html
- https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/comment%20out%20categories%20-%20MFA.png

I removed the `<br>` to no effect. [I tried many variations](https://github.com/JoelMTaylor/Documentation/tree/main/admin). For some reason, I didn't think to try that (maybe because it had an error the first time).

**Reply by geoffreymcgill** [reply]
All the `<br>` tags can be removed. Removing may not fix the rendering of the page, but nevertheless they should be removed. 

There does appear to be a problem with how Retype is handling the date parsing. In your `CorporateSystems/Multi-factor Authentication.md` document, you have an invalid `date` set to `date: 2021-03-1902`. This appears to be causing Retype to fail in a strange way and that might be causing the rendering the issue. 

https://github.com/JoelMTaylor/Documentation/blob/main/CorporateSystems/Multi-factor%20Authentication.md?plain=1#L7

Can you try setting to `date: 2021-03-19`?

**Reply by geoffreymcgill** [reply]
@JoelMTaylor The defect where configuring many `categories` caused a formatting issue has been fixed in the `v2.3.0` release which is now available. 

Hope this helps.

**Comment by geoffreymcgill** [comment]
If you have a frontmatter block, it will need to be the first thing on the `.md` page.

The following page will render, but not as you expect it to render as the frontmatter block is not the first item on the page.

https://github.com/JoelMTaylor/Documentation/blob/main/admin/Test%20Order%20Start3.md?plain=1#L2-L4

Also, those `<br>` tags should be removed. 

Hope this helps.

**Comment by geoffreymcgill** [comment]
A nice optimization that might work well for your project is moving each of the author configs into an include file. 

For example, if you create a `_includes/sara-dawson.md` file:

#### _includes/sara-dawson.md

```md
authors:
  - name: Sarah Dawson
    email:
    link:
    avatar: ../static/SarahDawson_Icon.png
```

and then just include that `sara-dawson.md` file into each page using `{{ include "sara-dawson" }}`, see:

#### sample.md
```
---
{{ include "sara-dawson" }}
categories:
  - Edumate Coordinators
---
!!!warning
Sample warning
!!!

# Heading Here

Text
```

This would allow you to manage just one instance of the author details for each author, and just include that chunk of configuration across many pages. Instead of reproducing and maintaining the exact same content on many individual pages. 

Hope this helps.

**Comment by geoffreymcgill** [comment, marked as answer]
Can you change the following line to use the `@latest` tag, instead of `@v2`?

https://github.com/JoelMTaylor/Documentation/blob/main/.github/workflows/retype-action.yml#L17

```yml
# Old
# - uses: retypeapp/action-build@v2

# New
- uses: retypeapp/action-build@latest
```

**Reply by JoelMTaylor** [reply]
I made all the changes that were suggested - thank you.

I was still having issues, so I decided to delete everything and then add them back in one/some at a time. Everything was fine until I added two files (MFA and troubleshooting). I reverted the addition. Since I moved files from one folder (outside the repo on my win ex), reverting the addition just deleted the files from the folder (did not move them back to where they came from). I had to recreate the files by copying the GitHub desktop logs (I probably could have copied them from my Documents repo which should have a duplicate). 

I moved MFA back in first - without issues. This leaves one file that could be the culprit (if a file or contents was the problem) - troubleshooting. In the GitHub logs, there was red circles being shown (something next to the text). 
![image](https://user-images.githubusercontent.com/97919641/162330861-071d6b3c-be58-4036-9184-fafbf8af548a.png)
I don't know what they are - they are not text to be copied - and I have no idea what they could be - there should be nothing there. My only guess is that whatever it was caused the issue. When recreating the troubleshooting file, I was careful not to copy it (even though copying the whole thing did not add any text in those locations - unknown or not). Adding the troubleshooting file back in did not cause any problems.

This was not the file set of files that had to be moved back. After moving some more files back, the pages became corrupt. This time, I did not permanently delete the files. Even though there were some red dots present, I could not see anything on the original files. 

I moved files in slowly to check any issues and any files that had red dots in them were recreated. I added all except for two files which were displaying the red dots. I tried the first (Creating New Users) and this caused the page corruption issues. I reverted and tried the second (Removing Users) which did not cause any issues. However, I reverted this, too, just to be safe. I have placed both files in the example files folder from previously along with a screenshot of the GitHub Log showing the red dots (note that the "🛇↵" always appear at the end).

However, this may not be the problem at all. I tried to recreate a 'safe' version of Creating New Users (it did not have any red dots in the GitHub log). But this file also caused issues. I reverted and added a blank file by the same name which still caused issues. I have changed the name to Creating Users and it had another issue. I removed the whole folder (including contents) and then re-added them back in and it seemed to work. 

The content was old and I needed to remove something from a few of the pages (an alert). However, when I did, the pages (all of them) were corrupted). I reverted, but it is still broken. 

I am probably going to try deleting them again to try get everything back to a working state. My only idea that I had about the root cause was that the troubleshooting page was changed recently (in the last day or two) and none/few other pages were changed. It is possible that something was added to that page (or it was somehow corrupted) and was causing all the issues.

Apart from that, I have no idea what is going on. I wrote all of this (and tried to be very methodical) so that I could isolate the issue for myself and the improvement of Retype (you could use my testing). It hasn't been much good to me apparently, but hopefully these notes will help you in some way to narrow down the problem do others don't experience it.

I will keep working on it and will hopefully get to a working state.

**Reply by geoffreymcgill** [reply]
Can you link me to the file that contains those characters?

**Reply by JoelMTaylor** [reply]
https://github.com/JoelMTaylor/Documentation/tree/main/_example%20files

**Reply by JoelMTaylor** [reply]
I just deleted everything (moved out) and then moved everything back in. Some pages remained corrupted, some got better (I don't think anything got worse). If you have any ideas that I could try, I'm open to any.

**Reply by geoffreymcgill** [reply]
Please link me to the `.md` source file that is creating the page with those characters in it.

**Reply by geoffreymcgill** [reply]
What `.md` file is being used to create the following page?

https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/authentication_Multi-factor%20authentication.html

At first it looks like the following page is being used, but the content is different between the `.md` and the `.html`, so you must be using a different version of the file that is not available in the repo. 

https://github.com/JoelMTaylor/Documentation/blob/main/CorporateSystems/Multi-factor%20Authentication.md

**Reply by JoelMTaylor** [reply]
> Please link me to the `.md` source file that is creating the page with those characters in it.

Those are:
- https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/Creating%20New%20Users.md
- https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/Removing%20Users.md

> What .md file is being used to create the following page?

That is part of the problem. The generated `.html` pages don't appear anything like the `.md` source. The pages should be the same.

**Reply by geoffreymcgill** [reply]
Well, I'm not sure what to say. The files you are linking me to are 100% completely different files:

The following `.html` file...
https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/authentication_Multi-factor%20authentication.html
... is not the result of building the following `.md` file:
https://github.com/JoelMTaylor/Documentation/blob/main/_example%20files/Creating%20New%20Users.md

I really don't know what is going on but the `.md` samples you are providing do not match up with the `.html` that your are posting.

Maybe it would be best to create a new clean repo and start again by just adding one file at time. There is something with the setup you are using that is just not reproducible in the GitHub repo. When I run the GitHub repo, the project appears to build and run correctly.

**Reply by JoelMTaylor** [reply]
@geoffreymcgill, you have been invited to the repository and you can access the repository website (which can be found in the reytpe.yml). 

Hopefully, this will be more useful in tracking down the issue. However, if the answer cannot be found, I have been considering restarting from scratch.

**Reply by geoffreymcgill** [reply]
Thank you. I was able to immediately reproduce the issue. There's something very strange going on with how the frontmatter block is being parsed and rendered. I'm working through a few tests to try and create a simplified sample that demonstrates how to reproduce the issue in an empty test project. 

I will keep you updated with my progress.

**Reply by geoffreymcgill** [reply]
We were able to isolate the issue to malformed `icon` configurations.

When using emoji shortcodes, the shortcode must be wrapped in double-quotes `"`. For instance, using `icon: ":telephone_receiver:"` instead of `icon: :telephone_receiver:`

```yml
# Existing
icon: :telephone_receiver:

# Revised
icon: ":telephone_receiver:"
```

I sent a pull-request with changes to three files with this `icon` formatting issue.

You will need to restart your instance of `retype watch`. First run `ctrl + c` to exit out of `retype watch` mode, then run the command `retype watch` again to restart the server. That should clear out the old bad formatting and things should be working better now.

Fixing those three `icon` configs does seem to fix the issue, but Retype should have handled those errors better. We are going to investigate and try to figure a better way to recover from those formatting errors.

I will keep this thread updated with our progress. At least for now, this work-around should at least fix the massive formatting issue.

**Reply by JoelMTaylor** [reply]
@geoffreymcgill, thank you for your assistance - we certainly wouldn't have been able to fix it without you.

**Reply by geoffreymcgill** [reply]
We were able to isolate the fix the defect that was causing the under lying formatting problems when the emoji shortcodes were unquoted. It was a very interesting problem, thanks for reporting. 

The fix will be available in the upcoming Retype `v2.3.0` release.

There's nothing different you will need to do in your app, but still continue to wrap emoji shortcodes in double-quotes, such as `":shortcode-here:"`. 

Hope this helps.
